rfd/0036/user-stories/jupiter-example-com-jenkins.md

```diff
@@ -0,0 +1,77 @@
+# Building and testing jupiter.example.com with Jenkins
+
+The trivial jupiter.example.com application consists of a number of stateless containers (or containers whose state can be easily reconstructed) running as a tiered application of Nginx fronting a PHP/Apache web content templating system.
+
+### Jenkins
+
+The operators use Jenkins to automate the process of building new containers for each pull request and branch in the origin repo. This requires that Jenkins jobs need to be able to interact with Mariposa via the CLI (or API).
+
+- Jenkins must be provisioned with an SSH key for Mariposa
+- Jenkins must be provisioned with an SSH key for GitHub deploy keys
+- Jenkins will effectively own the entire account until we have RBAC v2.
+- There needs to be a way to get the key onto Jenkins in the first place.
+
+Because Jenkins is going to build containers locally (via Docker on KVM), operators provision the SSH keys with traditional configuration management (ex. Ansible) and bind-mount this into the Jenkins container running on KVM. Optionally, the operators can treat the KVM as a traditional VM and deploy Jenkins to it via configuration management.
+
+### GitHub
+
+The code is hosted in GitHub. Most everything here varies only in detail with GitLab or Bitbucket. Each project (ex. jupiter.example.com) has a GitHub repository. The GitHub repository is configured with a webhook that sends push events to Jenkins. These push events include information on what branch and commit was pushed.
+
+### Jenkins jobs
+
+Jenkins should be configured with two jobs, one for development and one for production. The separate production job allows for manual approval of the build while reducing risk of accidental deployment (see below).
+
+### Jenkins development/test job
+
+This Jenkins job is configured to receive GitHub webhook pushes. The Jenkins job should run build and deployment tasks defined in the code repository so that the process for deployment is owned by the team that owns the code.
+
+Each GitHub branch has an associated Mariposa project. The Jenkins job clones a template manifest (from the repo) for each new branch. Each changeset received on a given branch will be deployed as a new version of that manifest.
```

Comment:
I imagine there's manifest per project (though we track changes in that manifest and can restore old versions).

I think my real point is that the passive voice there suggests that Jenkins isn't creating projects, so I wanted to be clear.